File: server/src/lib/storage.ts
import { S3Client, PutObjectCommand, DeleteObjectCommand, CreateBucketCommand, HeadBucketCommand } from "@aws-sdk/client-s3";
import { config } from "../config.js";
import { logger } from "./logger.js";

const isTest = config.NODE_ENV === "test";

let s3Client: S3Client | null = null;

if (!isTest) {
  s3Client = new S3Client({
    endpoint: config.S3_ENDPOINT,
    region: config.S3_REGION,
    credentials: {
      accessKeyId: config.S3_ACCESS_KEY_ID || "minioadmin",
      secretAccessKey: config.S3_SECRET_ACCESS_KEY || "minioadminpass",
    },
    forcePathStyle: true,
  });

  // Ensure bucket exists in MinIO asynchronously
  ensureBucketExists().catch((err) => {
    logger.warn({ err: err.name ?? err.message }, "MinIO bucket check failed. Ensure services are running.");
  });
}

async function ensureBucketExists() {
  if (!s3Client) return;
  const bucketName = config.S3_BUCKET_NAME || "intellihire";
  try {
    await s3Client.send(new HeadBucketCommand({ Bucket: bucketName }));
  } catch (err: any) {
    if (err.name === "NotFound" || err.$metadata?.httpStatusCode === 404) {
      logger.info({ bucketName }, "MinIO bucket not found. Creating bucket...");
      await s3Client.send(new CreateBucketCommand({ Bucket: bucketName }));
      logger.info({ bucketName }, "MinIO bucket created successfully");
    } else {
      throw err;
    }
  }
}

export async function uploadFile(key: string, body: Buffer, contentType: string): Promise<string> {
  const bucketName = config.S3_BUCKET_NAME || "intellihire";
  if (isTest || !s3Client) {
    // Return dummy URL during testing
    return `/dummy-storage/${bucketName}/${key}`;
  }

  await s3Client.send(
    new PutObjectCommand({
      Bucket: bucketName,
      Key: key,
      Body: body,
      ContentType: contentType,
    })
  );

  // Return the full public URL from the configured endpoint
  return `${config.S3_ENDPOINT}/${bucketName}/${key}`;
}

export async function deleteFile(key: string): Promise<void> {
  const bucketName = config.S3_BUCKET_NAME || "intellihire";
  if (isTest || !s3Client) return;

  await s3Client.send(
    new DeleteObjectCommand({
      Bucket: bucketName,
      Key: key,
    })
  );
}

export const storageService = {
  uploadFile,
  deleteFile,
};

File: server_py/app/storage.py
import os
import boto3
from botocore.exceptions import ClientError
from .logger import logger

NODE_ENV = os.getenv("NODE_ENV", "development")
is_test = NODE_ENV == "test"

# S3 Configuration
S3_ENDPOINT = os.getenv("S3_ENDPOINT", "http://localhost:9000")
S3_REGION = os.getenv("S3_REGION", "us-east-1")
S3_ACCESS_KEY_ID = os.getenv("S3_ACCESS_KEY_ID", "minioadmin")
S3_SECRET_ACCESS_KEY = os.getenv("S3_SECRET_ACCESS_KEY", "minioadminpass")
S3_BUCKET_NAME = os.getenv("S3_BUCKET_NAME", "intellihire")

s3_client = None

if not is_test:
    try:
        s3_client = boto3.client(
            "s3",
            endpoint_url=S3_ENDPOINT,
            region_name=S3_REGION,
            aws_access_key_id=S3_ACCESS_KEY_ID,
            aws_secret_access_key=S3_SECRET_ACCESS_KEY,
            config=boto3.session.Config(signature_version="s3v4")
        )
        
        # Ensure bucket exists
        try:
            s3_client.head_bucket(Bucket=S3_BUCKET_NAME)
        except ClientError as e:
            error_code = e.response.get("Error", {}).get("Code")
            if error_code in ("404", "NoSuchBucket"):
                logger.info(f"MinIO bucket {S3_BUCKET_NAME} not found. Creating bucket...")
                s3_client.create_bucket(Bucket=S3_BUCKET_NAME)
                logger.info(f"MinIO bucket {S3_BUCKET_NAME} created successfully")
            else:
                logger.warning(f"Error checking bucket: {e}")
    except Exception as err:
        logger.warning(f"MinIO storage initialization failed: {err}")

async def upload_file(key: str, body: bytes, content_type: str) -> str:
    bucket_name = S3_BUCKET_NAME
    if is_test or not s3_client:
        return f"/dummy-storage/{bucket_name}/{key}"

    try:
        s3_client.put_object(
            Bucket=bucket_name,
            Key=key,
            Body=body,
            ContentType=content_type
        )
        return f"{S3_ENDPOINT}/{bucket_name}/{key}"
    except Exception as e:
        logger.error(f"Failed to upload file to S3: {e}")
        raise e

async def delete_file(key: str) -> None:
    if is_test or not s3_client:
        return
    try:
        s3_client.delete_object(Bucket=S3_BUCKET_NAME, Key=key)
    except Exception as e:
        logger.error(f"Failed to delete file from S3: {e}")

storage_service = {
    "upload_file": upload_file,
    "delete_file": delete_file
}
